pkg/controller/mongodb/mongodb_controller.go

```diff
@@ -502,6 +603,40 @@ func buildStatefulSetModificationFunction(mdb mdbv1.MongoDB) statefulset.Modific
 
 	dataVolume := statefulset.CreateVolumeMount(dataVolumeName, "/data")
 
+	agentVolumeMounts := []corev1.VolumeMount{agentHealthStatusVolumeMount, automationConfigVolumeMount, dataVolume}
+	mongodVolumeMounts := []corev1.VolumeMount{mongodHealthStatusVolumeMount, hooksVolumeMount, dataVolume}
+
+	tlsPodSpec := podtemplatespec.Apply()
+	if mdb.Spec.TLS.Enabled {
+		// Configure an empty volume into which the TLS init container will write the certificate and key file
+		tlsVolume := statefulset.CreateVolumeFromEmptyDir("tls")
+		tlsVolumeMount := statefulset.CreateVolumeMount(tlsVolume.Name, tlsServerMountPath, statefulset.WithReadOnly(false))
+		agentVolumeMounts = append(agentVolumeMounts, tlsVolumeMount)
+		mongodVolumeMounts = append(mongodVolumeMounts, tlsVolumeMount)
+
+		// Configure a volume which mounts the CA certificate from a ConfigMap
+		// The certificate is used by both mongod and the agent
+		caVolume := statefulset.CreateVolumeFromConfigMap("tls-ca", mdb.Spec.TLS.CAConfigMapRef)
+		caVolumeMount := statefulset.CreateVolumeMount(caVolume.Name, tlsCAMountPath, statefulset.WithReadOnly(true))
+		agentVolumeMounts = append(agentVolumeMounts, caVolumeMount)
+		mongodVolumeMounts = append(mongodVolumeMounts, caVolumeMount)
+
+		// Configure a volume which mounts the secret holding the server key and certificate
+		// The same key-certificate pair is used for all servers
+		tlsSecretVolume := statefulset.CreateVolumeFromSecret("tls-secret", mdb.Spec.TLS.ServerSecretRef)
+		tlsSecretVolumeMount := statefulset.CreateVolumeMount(tlsSecretVolume.Name, tlsSecretMountPath, statefulset.WithReadOnly(true))
+
+		// MongoDB expects both key and certificate to be provided in a single PEM file
+		// We are using a secret format where they are stored in separate fields, tls.crt and tls.key
+		// Because of this we need to use an init container which reads the two files mounted from the secret and combines them into one
+		tlsPodSpec = podtemplatespec.Apply(
```

Comment:
very good candidate for a separate function to build this function I think

Reply:
yes the whole function is nearly 100 lines

Reply:
I've broken this out into a separate function now by introducing a new builder function: `podtemplatespec.WithVolumeMounts()`. The problem before was that the volume mounts needed to be applied to each container object which made it hard to refactor.